airbyte-integrations/connectors/source-sendgrid/source_sendgrid/config_migrations.py

```diff
@@ -0,0 +1,89 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+import logging
+from typing import Any, List, Mapping
+
+import pendulum
+from airbyte_cdk.config_observation import create_connector_config_control_message
+from airbyte_cdk.entrypoint import AirbyteEntrypoint
+from airbyte_cdk.sources import Source
+from airbyte_cdk.sources.message import InMemoryMessageRepository, MessageRepository
+
+logger = logging.getLogger("airbyte_logger")
+
+
+class MigrateToLowcodeConfig:
+    """
+    This class stands for migrating the config at runtime.
+    Specifically, starting from `0.5.0`, apikey moved to api_key and start_time moved to start_date.
+    start_time was not required and not start_date is.
+    """
+
+    message_repository: MessageRepository = InMemoryMessageRepository()
+
+    @classmethod
+    def should_migrate(cls, config: Mapping[str, Any]) -> bool:
+        """
+        This method determines if the config should be migrated.
+        Returns:
+            > True, if the transformation is necessary
+            > False, otherwise.
+        """
+        has_api = "apikey" in config
+        has_time = "start_time" in config
+        has_newapi = "api_key" in config
+        has_newdate = "start_date" in config
+        if has_newapi and has_newdate:
+            return False
+        return has_api or has_time
+
+    @staticmethod
+    def transform_config(config: Mapping[str, Any]) -> Mapping[str, Any]:
+        # move api_secret inside credentials block
+        if "apikey" in config:
+            config["api_key"] = config["apikey"]
+
+        if "start_time" in config:
+            config["start_date"] = pendulum.parse(config["start_time"]).to_iso8601_string()
+
+        if "start_date" not in config:
+            config["start_date"] = "2009-08-01T00:00:00Z"
```

Comment:
Would it be better to add this value as the default in the spec?

Reply:
I don't think so, we don't really _want_ people making a new source to pick the earliest possible date because it will take a long time to sync month by month from Sendgrid's creation.  However, I had to come up with a default for migration purposes.